undefined reference link error regression

This looks to be a regression from stable to beta unfortunately, compiling https://github.com/alexcrichton/futures-rs/commit/6ae0345ee0b2abc558b0d916fc17dbd46caa7969 succeeds on stable but fails on beta:

```
$ rustup run stable cargo build --manifest-path http/Cargo.toml
...
$ rustup run beta cargo build --manifest-path http/Cargo.toml
...
   Compiling http v0.1.0 (file:///home/alex/code/futures/http)
error: linking with `cc` failed: exit code: 1
note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/home/alex/code/futures/http/target/debug/http.0.o" "-o" "/home/alex/code/futures/http/target/debug/http" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/home/alex/code/futures/http/target/debug" "-L" "/home/alex/code/futures/http/target/debug/deps" "-L" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "/home/alex/code/futures/http/target/debug/libhttp.rlib" "/home/alex/code/futures/http/target/debug/deps/libhttparse-36b96d90ffbce562.rlib" "/home/alex/code/futures/http/target/debug/deps/libfuturemio-986c91cdeb491b00.rlib" "/home/alex/code/futures/http/target/debug/deps/libslab-45f4a226259c4869.rlib" "/home/alex/code/futures/http/target/debug/deps/libfutures-1a55edc1e7c129de.rlib" "/home/alex/code/futures/http/target/debug/deps/libmio-a89f84904d772c54.rlib" "/home/alex/code/futures/http/target/debug/deps/libnet2-b2390925f9bbf4a2.rlib" "/home/alex/code/futures/http/target/debug/deps/libcfg_if-72c1f992b13d5087.rlib" "/home/alex/code/futures/http/target/debug/deps/libmiow-b827bf6129dcd844.rlib" "/home/alex/code/futures/http/target/debug/deps/libwinapi-0889532d327ff4e2.rlib" "/home/alex/code/futures/http/target/debug/deps/libnix-f8d033ce6354dacb.rlib" "/home/alex/code/futures/http/target/debug/deps/libbytes-f9fbfc75eb2416e0.rlib" "/home/alex/code/futures/http/target/debug/deps/libscoped_tls-53036341d2dbbe26.rlib" "/home/alex/code/futures/http/target/debug/deps/libtime-750bfdd52feafcb7.rlib" "/home/alex/code/futures/http/target/debug/deps/liblibc-1f3392fe1afd1313.rlib" "/home/alex/code/futures/http/target/debug/deps/liblog-bf16bb9a4912b11d.rlib" "/home/alex/code/futures/http/target/debug/deps/libbitflags-e61ad67c3301e77d.rlib" "/home/alex/code/futures/http/target/debug/deps/libslab-4f8b9e9b6e35857c.rlib" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-411fd48b.rlib" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-411fd48b.rlib" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-411fd48b.rlib" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand-411fd48b.rlib" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcollections-411fd48b.rlib" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_unicode-411fd48b.rlib" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-411fd48b.rlib" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-411fd48b.rlib" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-411fd48b.rlib" "/home/alex/.multirust/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-411fd48b.rlib" "-l" "util" "-l" "dl" "-l" "pthread" "-l" "gcc_s" "-l" "pthread" "-l" "c" "-l" "m" "-l" "rt" "-l" "util" "-l" "compiler-rt"
note: /home/alex/code/futures/http/target/debug/http.0.o: In function `http::serve<http::request::Request,http::response::Response,closure>':
/home/alex/code/futures/http/src/lib.rs:55: undefined reference to `futuremio::tcp::_$LT$impl$u20$futuremio..event_loop..LoopHandle$GT$::tcp_listen::h9a7258751896fb30'
/home/alex/code/futures/http/target/debug/http.0.o: In function `http::serve::_$u7b$$u7b$closure$u7d$$u7d$::hed42a806f41e1837':
http.0.rs:(.text._ZN4http5serve28_$u7b$$u7b$closure$u7d$$u7d$17hed42a806f41e1837E+0xa2): undefined reference to `futuremio::tcp::TcpListener::incoming::h7aae24690b0d3a8d'
/home/alex/code/futures/http/target/debug/http.0.o: In function `futures::forget::forget<futures::then::Then<Box<Future>, core::result::Result<(), std::sync::mpsc::SendError<core::result::Result<(), std::io::error::Error>>>, closure>>':
/home/alex/code/futures/src/forget.rs:23: undefined reference to `futures::forget::_forget::h4301ad1ba69f01dc'
collect2: error: ld returned 1 exit status


error: aborting due to previous error
error: Could not compile `http`.
```

cc @michaelwoerister, does this look familiar?

I'll try to reduce down as well, but wanted to file an issue nonetheless.

Ok, minimized:

```rust
// foo.rs
#![crate_type = "lib"]
#![crate_name = "foo"]
pub trait Future {
    type Item;
    type Error;
}

fn foo() -> Box<Future<Item=(), Error=Box<()>>> {
    loop {}
}

pub fn bar<F, A, B>(_s: F)
    where F: Fn(A) -> B,
{
    foo();
}

// bar.rs
extern crate foo;

fn mk<T>() -> T { loop {} }

struct Data<T, E> {
    data: T,
    error: E,
}

fn main() {
    foo::bar(|()| {
        Data::<(), std::io::Error> {
            data: mk(),
            error: mk(),
        }
    })
}
```

```
$ rustup run stable rustc foo.rs && rustup run stable rustc bar.rs -L .
...
$ rustup run nightly rustc foo.rs && rustup run nightly rustc bar.rs -L .
...
error: linking with `cc` failed: exit code: 1
note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "bar.0.o" "-o" "bar" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "." "-L" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "/home/alex/code/futures/libfoo.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-c8005792.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-c8005792.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-c8005792.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand-c8005792.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcollections-c8005792.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_unicode-c8005792.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-c8005792.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-c8005792.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-c8005792.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-c8005792.rlib" "-l" "dl" "-l" "pthread" "-l" "gcc_s" "-l" "pthread" "-l" "c" "-l" "m" "-l" "rt" "-l" "util" "-l" "compiler-rt"
note: bar.0.o: In function `foo::bar::h97da9529975f6f18':
bar.cgu-0.rs:(.text._ZN3foo3bar17h97da9529975f6f18E+0xf): undefined reference to `foo::foo::h4e3fa806e0391cb9'
collect2: error: ld returned 1 exit status
```

--- a/src/librustc_metadata/tyencode.rs
+++ b/src/librustc_metadata/tyencode.rs
@@ -407,7 +407,14 @@ pub fn enc_existential_bounds<'a,'tcx>(w: &mut Cursor<Vec<u8>>,
 
     enc_region(w, cx, bs.region_bound);
 
-    for tp in &bs.projection_bounds {
+    // Encode projection_bounds in a stable order
+    let mut projection_bounds: Vec<_> = bs.projection_bounds
+                                          .iter()
+                                          .map(|b| (b.item_name().as_str(), b))
+                                          .collect();
+    projection_bounds.sort_by_key(|&(ref name, _)| name.clone());
+
+    for tp in projection_bounds.iter().map(|&(_, tp)| tp) {
         write!(w, "P");
         enc_projection_predicate(w, cx, &tp.0);
     }

--- a/src/test/run-pass/auxiliary/issue34796aux.rs
+++ b/src/test/run-pass/auxiliary/issue34796aux.rs
@@ -1,0 +1,30 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![crate_type = "lib"]
+pub trait Future {
+    type Item;
+    type Error;
+}
+
+impl Future for u32 {
+    type Item = ();
+    type Error = Box<()>;
+}
+
+fn foo() -> Box<Future<Item=(), Error=Box<()>>> {
+    Box::new(0u32)
+}
+
+pub fn bar<F, A, B>(_s: F)
+    where F: Fn(A) -> B,
+{
+    foo();
+}

--- a/src/test/run-pass/issue34796.rs
+++ b/src/test/run-pass/issue34796.rs
@@ -1,0 +1,36 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// This test case exposes conditions where the encoding of a trait object type
+// with projection predicates would differ between this crate and the upstream
+// crate, because the predicates were encoded in different order within each
+// crate. This led to different symbol hashes of functions using these type,
+// which in turn led to linker errors because the two crates would not agree on
+// the symbol name.
+// The fix was to make the order in which predicates get encoded stable.
+
+// aux-build:issue34796aux.rs
+extern crate issue34796aux;
+
+fn mk<T>() -> T { loop {} }
+
+struct Data<T, E> {
+    data: T,
+    error: E,
+}
+
+fn main() {
+    issue34796aux::bar(|()| {
+        Data::<(), std::io::Error> {
+            data: mk(),
+            error: mk(),
+        }
+    })
+}

